### ISSUE TITLE
[major change] MIR librarification

### ISSUE THREAD

**Issue by oli-obk** [rust-lang/compiler-team#233]
# TLDR

Move the MIR datastructures into a separate crate that does not depend on `librustc`. The dependency on `Ty<'tcx>` is broken by introducing a generic parameter.

# Links and Details

MVP (just a single type that depends on `Ty<'tcx>` and one that does not): https://github.com/rust-lang/rust/compare/master...oli-obk:mir_librarification?expand=1

Zulip discussion: https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/MIR.20librarification/near/183943342

# Mentors or Reviewers

* @Centril (for the splitting)
* @nikomatsakis (for later adding a dependency on chalk-ir so some operations that need some info about types can be moved to the new crate, too)

**Comment by Centril** [comment]
I suspect I'm not the best reviewer there is for this since I've done only minor work on `Ty` and MIR, but the adding of a type parameter seems fairly straightforward so I could review.

Some elaboration on the expected gains & drawbacks from this change could be good tho.

**Comment by oli-obk** [comment]
Drawbacks:
* more generics, may increase build time.
* more generics, may make it harder to add new logic to `rustc_mir_base`

Gains: 
* Can use MIR without requiring `TyCtxt` or other rustc specific stuff
* I'm hoping for a cleaner API because we actually have to think about it instead of just being able to entangle it with all the rustc datastructures.
* Less stuff in librustc

**Comment by Centril** [comment]
> more generics, may increase build time.

> Less stuff in librustc

(These may hopefully cancel each other out thanks to pipelining.)

**Comment by spastorino** [comment]
Added to our compiler team meeting agenda, removing `to-announce` label.

**Comment by nikomatsakis** [comment]
@oli-obk and I discussed briefly and decided that it would make sense to wait on this until we have actual consumers of MIR, and/or until library-ification has proceeded further.

For example, if we have a library for rust types generic over an `I: Interner`, then maybe we would want MIR to build on that be generic over an `I: Interner` itself?
